crates/primitives/src/signature/sig.rs

```diff
@@ -0,0 +1,727 @@
+    }
+
+    /// Recovers an [`Address`] from this signature and the given prehashed message.
+    ///
+    /// [`Address`]: crate::Address
+    #[inline]
+    pub fn recover_address_from_prehash(
+        &self,
+        prehash: crate::B256,
+    ) -> Result<crate::Address, SignatureError> {
+        self.recover_from_prehash(prehash).map(|vk| crate::Address::from_public_key(&vk))
+    }
+
+    /// Recovers a [`VerifyingKey`] from this signature and the given message by first prefixing and
+    /// hashing the message according to [EIP-191](crate::eip191_hash_message).
+    ///
+    /// [`VerifyingKey`]: k256::ecdsa::VerifyingKey
+    #[inline]
+    pub fn recover_from_msg<T: AsRef<[u8]>>(
+        &self,
+        msg: T,
+    ) -> Result<k256::ecdsa::VerifyingKey, SignatureError> {
+        self.recover_from_prehash(crate::eip191_hash_message(msg))
+    }
+
+    /// Recovers a [`VerifyingKey`] from this signature and the given prehashed message.
+    ///
+    /// [`VerifyingKey`]: k256::ecdsa::VerifyingKey
+    #[inline]
+    pub fn recover_from_prehash(
+        &self,
+        prehash: crate::B256,
+    ) -> Result<k256::ecdsa::VerifyingKey, SignatureError> {
+        let this = self.normalize_s().unwrap_or(*self);
+        k256::ecdsa::VerifyingKey::recover_from_prehash(
+            prehash.as_slice(),
+            &this.inner,
+            this.recid(),
+        )
+        .map_err(Into::into)
+    }
+}
+
+impl Signature<()> {
+    /// Parses a signature from a byte slice, with a v value
+    ///
+    /// # Panics
+    ///
+    /// If the slice is not at least 64 bytes long.
+    #[inline]
+    pub fn from_bytes_and_parity<T: TryInto<Parity, Error = E>, E: Into<SignatureError>>(
+        bytes: &[u8],
+        parity: T,
+    ) -> Result<Self, SignatureError> {
+        let r = U256::from_be_slice(&bytes[..32]);
+        let s = U256::from_be_slice(&bytes[32..64]);
+        Self::from_rs_and_parity(r, s, parity)
+    }
+
+    /// Instantiate from v, r, s.
+    pub fn from_rs_and_parity<T: TryInto<Parity, Error = E>, E: Into<SignatureError>>(
+        r: U256,
+        s: U256,
+        parity: T,
+    ) -> Result<Self, SignatureError> {
+        Ok(Self { inner: (), v: parity.try_into().map_err(Into::into)?, r, s })
+    }
+}
+
+impl<S: Copy> Signature<S> {
+    /// Returns the inner ECDSA signature.
+    #[inline]
+    pub const fn into_inner(self) -> S {
+        self.inner
+    }
+
+    /// Modifies the recovery ID by applying [EIP-155] to a `v` value.
+    ///
+    /// [EIP-155]: https://eips.ethereum.org/EIPS/eip-155
+    #[inline]
+    pub fn with_chain_id(self, chain_id: u64) -> Self {
+        self.with_parity(self.v.with_chain_id(chain_id))
+    }
+
+    /// Modifies the recovery ID by dropping any [EIP-155] v value, converting
+    /// to a simple parity bool.
+    pub fn with_parity_bool(self) -> Self {
+        self.with_parity(self.v.to_parity_bool())
+    }
+}
+
+impl<S> Signature<S> {
```

Comment:
conversions for `Signature<_> -> Signature<k256>`?

Reply:
again, because we're only exporting a parameterized version, we don't need any conversions. no program will be using multiple `Signature<T>` with different T

Reply:
What about the existing `<()>`?

Reply:
no downstream user can instantiate anything _other than_ `crate::Signature`

Reply:
Ah right though both () and k256 were exposed, but there's only one feature-gated export